File: src/08.py
import sys

def parse_input():
    lines = sys.stdin.read().strip().split('\n')
    grid = [list(line.strip()) for line in lines]
    return grid

def find_antinodes(grid):
    antinode_positions = set()
    antinode_positions_with_harmonics = set()
    freq_to_positions = {}

    # Populate the frequency dictionary with antenna positions
    rows, cols = len(grid), len(grid[0])
    for i in range(rows):
        for j in range(cols):
            char = grid[i][j]
            if char != '.':
                if char not in freq_to_positions:
                    freq_to_positions[char] = []
                freq_to_positions[char].append((i, j))

    # Calculate antinodes for each frequency
    for freq, positions in freq_to_positions.items():
        for i in range(len(positions)):
            x1, y1 = positions[i]
            for j in range(i + 1, len(positions)):
                x2, y2 = positions[j]
                
                # Calculate differences
                dx, dy = x2 - x1, y2 - y1

                # Check if one is twice as far away in any direction
                antinode_pos1 = (x1 - dx, y1 - dy)
                antinode_pos2 = (x2 + dx, y2 + dy)
                
                if 0 <= antinode_pos1[0] < rows and 0 <= antinode_pos1[1] < cols:
                    antinode_positions.add(antinode_pos1)
                if 0 <= antinode_pos2[0] < rows and 0 <= antinode_pos2[1] < cols:
                    antinode_positions.add(antinode_pos2)

                antinode_pos = (x1, y1)
                while 0 <= antinode_pos[0] < rows and 0 <= antinode_pos[1] < cols:
                    antinode_positions_with_harmonics.add(antinode_pos)
                    antinode_pos = (antinode_pos[0] - dx, antinode_pos[1] - dy)
                
                antinode_pos = (x2, y2)
                while 0 <= antinode_pos[0] < rows and 0 <= antinode_pos[1] < cols:
                    antinode_positions_with_harmonics.add(antinode_pos)
                    antinode_pos = (antinode_pos[0] + dx, antinode_pos[1] + dy)

    return antinode_positions, antinode_positions_with_harmonics

def main():
    grid = parse_input()
    part1, part2 = find_antinodes(grid)
    print(len(part1))
    print(len(part2))

if __name__ == "__main__":
    main()

File: src/08.rs
use std::collections::{HashSet, HashMap};
use std::io::{self, Read}; // Add this line

fn parse_input() -> Vec<Vec<char>> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    input.trim().lines()
        .map(|line| line.chars().collect())
        .collect()
}

fn find_antinodes(grid: &Vec<Vec<char>>) -> (HashSet<(usize, usize)>, HashSet<(usize, usize)>) {
    let mut antinode_positions = HashSet::new();
    let mut antinode_positions_with_harmonics = HashSet::new();
    let mut freq_to_positions = HashMap::new();

    // Populate the frequency dictionary with antenna positions
    let rows = grid.len();
    let cols = grid[0].len();
    for i in 0..rows {
        for j in 0..cols {
            let char = grid[i][j];
            if char != '.' {
                freq_to_positions.entry(char).or_insert(vec![]).push((i, j));
            }
        }
    }

    // Calculate antinodes for each frequency
    for positions in freq_to_positions.values() {
        for i in 0..positions.len() {
            let (x1, y1) = positions[i];
            for j in i + 1..positions.len() {
                let (x2, y2) = positions[j];

                // Calculate differences
                let dx = x2 - x1;
                let dy = y2 - y1;

                // Check if one is twice as far away in any direction
                let antinode_pos1 = (x1 - dx, y1 - dy);
                let antinode_pos2 = (x2 + dx, y2 + dy);

                if antinode_pos1.0 < rows && antinode_pos1.1 < cols {
                    antinode_positions.insert(antinode_pos1);
                }
                if antinode_pos2.0 < rows && antinode_pos2.1 < cols {
                    antinode_positions.insert(antinode_pos2);
                }

                let mut antinode_pos = (x1, y1);
                while antinode_pos.0 < rows && antinode_pos.1 < cols {
                    antinode_positions_with_harmonics.insert(antinode_pos);
                    antinode_pos = (antinode_pos.0 - dx, antinode_pos.1 - dy);
                }

                let mut antinode_pos = (x2, y2);
                while antinode_pos.0 < rows && antinode_pos.1 < cols {
                    antinode_positions_with_harmonics.insert(antinode_pos);
                    antinode_pos = (antinode_pos.0 + dx, antinode_pos.1 + dy);
                }
            }
        }
    }

    (antinode_positions, antinode_positions_with_harmonics)
}

fn main() {
    let grid = parse_input();
    let (part1, part2) = find_antinodes(&grid);
    println!("{}", part1.len());
    println!("{}", part2.len());
}